File: sdk/scala-sdk-protobuf-testkit/src/main/scala/kalix/scalasdk/testkit/MockRegistry.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk.testkit

import kalix.scalasdk.testkit.impl.MockRegistryImpl

import scala.reflect.ClassTag

/**
 * This trait is meant to allow for unit testing when a service has cross-component or cross-service calls. The set of
 * mocks or stubs will be matched by its class type upon a call of an external component or service.
 */
trait MockRegistry {

  /**
   * Returns a new MockRegistry with the new mock added to previous ones.
   *
   * @param instance
   *   The instance object to be used as a mock.
   * @param expectedClass
   *   The class type of the object being mocked.
   * @tparam T
   *   The service interface to be mocked.
   * @return
   *   A copy of this MockRegistry.
   */
  def withMock[T](instance: T)(implicit expectedClass: ClassTag[T]): MockRegistry
}

object MockRegistry {
  val empty = new MockRegistryImpl()

  /**
   * Returns a new MockRegistry with the new mock added to previous ones.
   *
   * @param instance
   *   The instance object to be used as a mock.
   * @param expectedClass
   *   The class type of the object being mocked.
   * @tparam T
   *   The service interface to be mocked.
   * @return
   *   A copy of this MockRegistry.
   */
  def withMock[T](instance: T)(implicit expectedClass: ClassTag[T]): MockRegistry =
    new MockRegistryImpl(Map(expectedClass.runtimeClass -> instance))
}

File: sdk/java-sdk-protobuf-testkit/src/main/java/kalix/javasdk/testkit/MockRegistry.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.testkit;

import kalix.javasdk.testkit.impl.MockRegistryImpl;

/**
 * This trait is meant to allow for unit testing when a service has cross-component or cross-service
 * calls. The set of mocks or stubs will be matched by its class type upon a call of an external
 * component or service.
 */
public interface MockRegistry {
  /**
   * Returns a new MockRegistry with the new mock added to previous ones.
   *
   * @param clazz The class type used to identify the mock.
   * @param instance The instance object to be used as a mock-
   * @return A copy of this MockRegistry.
   * @param <T> The service interface to be mocked.
   */
  <T> MockRegistry withMock(Class<T> clazz, T instance);

  /**
   * Returns an empty instance of MockRegistry that can be chained with `withMock`
   *
   * @return a new instance of MockRegistry
   */
  static MockRegistry create() {
    return MockRegistryImpl.empty();
  }

  MockRegistry EMPTY = MockRegistryImpl.empty();
}
